File: Fundamentals_of_Software_Design/1_Basic_skills/Стохастическая оценка числа Pi/main.py
import random

def estimate_pi(num_points):
    inside_circle = 0

    for _ in range(num_points):
        # Генерируем случайные координаты (x, y) в диапазоне [0, 1]
        x = random.uniform(0, 1)
        y = random.uniform(0, 1)
        
        # Проверяем, попадает ли точка в четверть окружности
        if x**2 + y**2 <= 1:
            inside_circle += 1

    # Площадь четверти круга приближенно равна отношению точек внутри круга к общему числу точек
    # Умножаем на 4 для получения площади полного круга
    return (inside_circle / num_points) * 4

# Ввод количества точек
num_points = int(input("Введите количество точек для метода Монте-Карло: "))

# Оценка числа π
pi_estimate = estimate_pi(num_points)
print(f"Оценка числа π: {pi_estimate}")

File: Fundamentals_of_Software_Design/1_Basic_skills/Стохастическая оценка числа Pi/main.c
#include <stdio.h>
#include <stdlib.h>
#include <time.h>
#include <math.h>

double estimate_pi(int num_points) {
    int inside_circle = 0;

    for (int i = 0; i < num_points; i++) {
        // Генерация случайных координат (x, y) в диапазоне [0, 1]
        double x = (double)rand() / RAND_MAX;
        double y = (double)rand() / RAND_MAX;

        // Проверяем, попадает ли точка в четверть окружности
        if (x * x + y * y <= 1.0) {
            inside_circle++;
        }
    }

    // Вычисление числа π
    return (4.0 * inside_circle) / num_points;
}

int main() {
    int num_points;

    // Инициализация генератора случайных чисел
    srand((unsigned int)time(NULL));

    // Ввод количества точек
    printf("Введите количество точек для метода Монте-Карло: ");
    scanf("%d", &num_points);

    if (num_points <= 0) {
        printf("Количество точек должно быть положительным.\n");
        return 1;
    }

    // Оценка числа π
    double pi_estimate = estimate_pi(num_points);
    printf("Оценка числа π: %.6f\n", pi_estimate);

    return 0;
}
